#' Plot protein features.
#'
#' @param res proteinFeatures
#' @param traces.obj An object of type \code{traces.obj}.
#' @param proteinID protein ID
#' @param plot_peak logical
#' @param plot_monomer logical
#' @param log logical
#' @param plot_apex logical
#' @param plot_in_complex_estimate logical
#' @param legend logical
#' @param highlight vector of trace ids to highlight
#' @param showtraces character Either "annotated" or "detected".
#' @param plot boolean, wether to print the plot or return the plot object
#' @param 
#' @export

plot.proteinFeatures <- function(res,
                                 traces.obj,
                                 proteinID,
                                 plot_peak=TRUE,
                                 plot_monomer=TRUE,
                                 log=FALSE,
                                 plot_apex=TRUE,
                                 plot_in_complex_estimate=TRUE,
                                 legend=FALSE,
                                 highlight=NULL,
                                 showtraces="annotated",
                                 plot = TRUE) {


  features <- subset(res, protein_id == proteinID)
  if(showtraces == "annotated"){
    peptides <- unique(unlist(strsplit(features$subunits_annotated, split = ";")))
  }else if(showtraces == "detected"){
    peptides <- unique(unlist(strsplit(features$subunits_detected, split = ";")))
  }else if(showtraces == "all"){
    peptides <- traces.obj$trace_annotation[protein_id == proteinID, id]
  }else{
    stop("Parameter 'showtraces' must be either 'annotated', 'detected', or 'all'")
  }
  # peptides <- traces.obj$trace_annotation[protein_id == proteinID]$id
  traces <- subset(traces.obj,trace_ids = peptides)
  traces.long <- toLongFormat(traces$traces)
  if(!is.null(highlight)){
    traces.long$outlier <- traces.long$id %in% highlight
  }
  # add monomer MW
  monomer = as.numeric(unlist(features$monomer_sec))
  monomer_max = as.numeric(unlist(features$monomer_sec_max))

  setkey(traces.long, id)

  proteinName = unique(traces$trace_annotation$protein_id)[1]

  group.colors <- c("TRUE" = "green3", "FALSE" = "firebrick1")

  p <- ggplot(traces.long) +
    geom_line(aes_string(x='fraction', y='intensity', color='id')) +
    ggtitle(proteinName) +
    xlab('fraction') +
    ylab('intensity')
  
    if(!legend){
      p <- p + guides(color=FALSE)
    }
    
    if (log) {
      p <- p + scale_y_log10('log(intensity)')
    }
  if(nrow(features) >= 1){
    if(plot_peak==TRUE){
      if (plot_in_complex_estimate == TRUE) {
        p <- p + geom_rect(data=features,aes(xmin = left_pp, xmax = right_pp, ymin = 0, ymax = Inf, fill=in_complex),alpha = 0.25) +
          scale_fill_manual(values=group.colors)
      } else {
        p <- p + geom_rect(data=features,aes(xmin = left_pp, xmax = right_pp, ymin = 0, ymax = Inf),alpha = 0.25)
      }
      p <- p + geom_vline(data=features,aes(xintercept = left_pp), colour="darkgrey", linetype="dashed")
      p <- p + geom_vline(data=features,aes(xintercept = right_pp), colour="darkgrey",linetype="dashed")
    }
    
    if (plot_monomer==TRUE){
      p <- p + geom_vline(xintercept=monomer,linetype="solid")
      p <- p + geom_vline(xintercept=monomer_max,linetype="dashed")
    }
    if(plot_apex==TRUE){
      p <- p + geom_vline(data=features,aes(xintercept=apex), colour="darkgrey", linetype="solid")
    }
  }
  #p <- p + theme(legend.position="none")
  if(!is.null(highlight)){
    legend_peps <- unique(traces.long[outlier == TRUE, id])
  p <- p + 
    geom_line(data = traces.long[outlier == TRUE], aes_string(x='fraction', y='intensity', color='id'), lwd=2) +
    scale_color_discrete(breaks = legend_peps)
  }
  if(plot){
    print(p)
  }else{
    return(p)
  }
  
}


#' Plot protein features with peptide alignment to sequence.
#'
#' @param res proteinFeatures
#' @param traces.obj An object of type \code{traces.obj}.
#' @param proteinID protein ID
#' @param ProteinFasta A path to a fasta file with protein sequences or an \code{AAStringSet} object.
#' @param outfile_name Name of the pdf to be printed
#' @param pdf \code{boolean} Wether to print a pdf file or to console
#' @param highlight vector of trace ids to highlight
#' @param 
#' @export
#' 
#' @param proteinIDs Character vector of proteins to plot, 
#' matching with names in the protein_id column of the features table.
#' @param traces An object of type \code{traces.obj}.
#' @param features An object of type \code{complexFeatures} as generated by the 
#'  findComplexFeatures function.
#' @param highlight_cutoff Traces with this combined_diff score or higher are highlighted.
#' @param outfile_name Name of the pdf to be printed
#' @param pdf \code{boolean} Wether to print a pdf file or to console
#' @param diff_scores list of diff_scores produced by calculateFeatureDiffScores
#' @param plot_sequence \code{boolean} Wether to also print the protein sequence
#' This is ignored if \code{plot_sequence} is set to \code{FALSE} and reqired if \code{TRUE}.
#' @import Biostrings
#' @import msa
#' @export

plot.proteinFeaturesSequence <- function(res,
                                         traces.obj,
                                         proteinID,
                                         ProteinFasta,
                                         highlight=NULL,
                                         outfile_name = NULL) {
  if(class(ProteinFasta) == "character"){
    sequences <- readAAStringSet(filepath = ProteinFasta)
  }else if(class(ProteinFasta) == "AAStringSet"){
    sequences <- ProteinFasta
  }else{
    stop("Input ProteinFasta must be of type character (Filepath to fasta file) or AAStringSet")
  }
  # if(grepl("^>ENSG", names(sequences[1]))){
  #   names(sequences) <- gsub("^>|-.*|\\|.*","", names(sequences))
  # }else{
  #   names(sequences) <- gsubsapply(strsplit(names(sequences), split = "\\|"), "[", 3)
  # }
  if(is.null(outfile_name)) outfile_name <- paste0("ProteinFeatures_", as.character(proteinID), collapse = "")
  if(!grepl("\\.pdf$", outfile_name)) outfile_name <- paste0(outfile_name,".pdf")
  # Pass the peptides to highlight
  top <- data.frame(rep(1, length(highlight)))
  rownames(top) <- highlight
  diff_scores <- NULL
  features <- res
  traces <- traces.obj
  proteinIDs <- proteinID
  rmarkdown::render(input = base::system.file("R/plotProteinFeaturesSequence.Rmd",package = "SECprofiler"),
                    output_file = outfile_name, output_dir = getwd(), knit_root_dir = getwd())
  
  
}


