# Functions for using SECprofiler with a dataset generated by a custom sequence Database

#' Extract Peptides that are Isoform specific
#' @import data.table
#' @param peptideTable Quantitative MS data in form of OpenSWATH result file or R data.table.
#' @return A named vector that conatins all fullPeptideNames of the input file that are isoform specific.
#' The Names correspond to the ProteinNames in the input Table
#' @export

getIsoPeptides <- function(peptideTable){
  protein_names <- lapply(strsplit(peptideTable$ProteinName, split = "/"),"[", -1)
  isIsoformspecific <- sapply(protein_names, function(x) all(grepl("_",x)))
  isopeps <- peptideTable[isIsoformspecific, FullPeptideName]
  names(isopeps) <- peptideTable[isIsoformspecific, ProteinName]
  isopeps <- isopeps[!duplicated(isopeps)]
  return(isopeps)
}

#' Filter an OpenSWATH output for Peptides that are Isoform specific
#' @import data.table
#' @param peptideTable Quantitative MS data in form of OpenSWATH result file or R data.table.
#' @return An isoform-peptide filtered peptideTable.
#' @export

filterIsoPeptides <- function(peptideTable){
  protein_names <- lapply(strsplit(peptideTable$ProteinName, split = "/"),"[", -1)
  isIsoformspecific <- sapply(protein_names, function(x) all(grepl("_",x)))
  peptideTable <- peptideTable[isIsoformspecific]
  return(peptideTable)
}


#' Change the protein names in the Open Swath output from custom DB search to make the import Isoform aware
#' @import data.table
#' @import dplyr
#' @param openSWATH Quantitative MS data in form of OpenSWATH result file or R data.table.
#' @param header_mapping Mapping table that is produced with the header-mapping function
#'  from the RNAseqToCustomDB package.
#' @return A vector with geneNames that can be used to substitute the ProteinName column of the OpenSWATH output table.
#' @export

mapProteinNamesToGeneLevel <- function(openSWATH, header_mapping){
  split <- strsplit(openSWATH$ProteinName, split = "/")
  header_mapping[,new_id := paste0(gene, "-", Isoform)]
  iRT <- readRDS("iRT.rda")
  iRT$new_id <- gsub(">| .*", "", iRT$header)
  iRT$protein <- gsub(">| .*", "", iRT$header)
  cRAP_tags <- readRDS("cRAP_tags.rda")
  cRAP_tags$new_id <- gsub(">| .*", "", cRAP_tags$header)
  cRAP_tags$protein <- gsub(">| .*", "", cRAP_tags$header)
  # library(dplyr)
  header_mapping <- bind_rows(header_mapping, iRT)
  header_mapping <- bind_rows(header_mapping, cRAP_tags)
  setkey(header_mapping,protein)
  newprot <- sapply(split, mapSinglePeptide, header_mapping)
  return(newprot)
}

mapSinglePeptide <- function(proteins,header_mapping){
  newprot <- header_mapping[proteins[-1], nomatch = 0, new_id]
  # Do this to mimic swossprot header format
  newprot <- paste0("cf|", newprot, "|peptide")
  # proteins[2:length(proteins)] <- sapply(proteins[2:length(proteins)], function(x){
  #   header_mapping[protein == x, paste0(gene, "-", Isoform)]
  # })
  # print(1)
  return(paste(c(proteins[1],newprot), collapse = "/"))
}
